jira.go

```diff
@@ -267,9 +289,75 @@ func (c *Client) NewMultiPartRequest(method, urlStr string, buf *bytes.Buffer) (
 	return c.NewMultiPartRequestWithContext(context.Background(), method, urlStr, buf)
 }
 
+func (c *Client) Do(req *http.Request, v interface{}) (*Response, error) {
+	return c.doWithDeadline(req, v)
+}
+
+func (c *Client) doWithDeadline(req *http.Request, v interface{}) (*Response, error) {
+	ctx := req.Context()
+	if failFast(ctx) {
+		return c.do(req, v)
+	}
+
+	deadline, deadlineIsSet := ctx.Deadline()
+	if !deadlineIsSet {
+		deadline = time.Now().Add(retryIntervalDefault)
+		dctx, cancel := context.WithDeadline(ctx, deadline)
+		defer cancel()
+		ctx = dctx
+		req = req.Clone(ctx)
+	}
+	errDeadlineWrapped := func(resp *Response, err error) error {
+		return errors.Wrapf(context.DeadlineExceeded, "path: %s, err: %+v, code: %d", req.URL.String(), err, resp.StatusCode)
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		default:
+		}
+
+		resp, err := c.do(req, v)
+		if resp != nil {
+			if _, ok := retryStatusCodes[resp.StatusCode]; !ok {
+				return resp, err
+			}
+			retryAfter := resp.Header.Get("Retry-After")
+			if retryAfter == "" {
+				t := time.NewTimer(retryStep)
+				select {
+				case <-ctx.Done():
+					t.Stop()
+					return nil, errDeadlineWrapped(resp, err)
+				case <-t.C:
+				}
+				continue
+			}
+			retryAfterSeconds, errConv := strconv.Atoi(retryAfter)
+			if errConv != nil {
+				return resp, errors.Wrap(err, errConv.Error())
```

Comment:
Are you sure you want to fail because of malformed header, which you allow to be empty? I think we should at most log it, but pretend that it was empty in case of an error.

Reply:
Done.